pkg/agent/cniserver/interface_configuration_windows.go

```diff
@@ -181,26 +223,87 @@ func (ic *ifConfigurator) createContainerLink(endpointName string, result *curre
 	return hnsEP, nil
 }
 
+func (ic *ifConfigurator) getInterfaceConfigForPods(pods sets.String) map[string]*interfacestore.InterfaceConfig {
+	interfaces := make(map[string]*interfacestore.InterfaceConfig)
+	ic.epCache.Range(func(key, value interface{}) bool {
+		ep, _ := value.(*hcsshim.HNSEndpoint)
+		ifConfig := parseOVSPortInterfaceConfigFromHNSEndpoint(ep)
+		if ifConfig == nil {
+			return true
+		}
+		namespacedName := k8s.NamespacedName(ifConfig.PodNamespace, ifConfig.PodName)
+		if pods.Has(namespacedName) {
+			interfaces[namespacedName] = ifConfig
+		}
+		return true
+	})
+	return interfaces
+}
+
+func parseOVSPortInterfaceConfigFromHNSEndpoint(ep *hcsshim.HNSEndpoint) *interfacestore.InterfaceConfig {
+	portData := &ovsconfig.OVSPortData{
+		Name:        ep.Name,
+		ExternalIDs: ep.AdditionalParams,
+	}
+	ifaceConfig := ParseOVSPortInterfaceConfig(portData, nil)
+	if ifaceConfig != nil {
+		var err error
+		ifaceConfig.MAC, err = net.ParseMAC(ep.MacAddress)
+		if err != nil {
+			klog.Errorf("Failed to parse MAC address from HNSEndpoint %s: %v", ep.MacAddress, err)
+			return nil
+		}
+	}
+	return ifaceConfig
+}
+
 // attachContainerLink takes the result of the IPAM plugin, and adds the appropriate IP
-// addresses and routes to the interface. It then sends a gratuitous ARP to the network.
+// addresses and routes to the interface.
+// For different CRI runtime we need to use corresponding Windows container API:
+//   - Docker runtime: HNS API
+//   - ContainerD runtime: HCS API
 func attachContainerLink(ep *hcsshim.HNSEndpoint, containerID, sandbox, containerIFDev string) (*current.Interface, error) {
 	var attached bool
-	attached, err := ep.IsAttached(containerID)
-	if err != nil {
-		return nil, err
+	var err error
+	var hcnEp *hcn.HostComputeEndpoint
+	if isDockerContainer(sandbox) {
+		attached, err = ep.IsAttached(containerID)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		if hcnEp, err = hcn.GetEndpointByID(ep.Id); err != nil {
+			return nil, err
+		}
+		attachedEpIds, err := hcn.GetNamespaceEndpointIds(sandbox)
+		if err != nil {
+			return nil, err
+		}
+		for _, existingEP := range attachedEpIds {
+			if existingEP == hcnEp.Id {
+				attached = true
+				break
+			}
+		}
 	}
 
 	if attached {
 		klog.V(2).Infof("HNS Endpoint %s already attached on container %s", ep.Id, containerID)
 	} else {
-		if err := hcsshim.HotAttachEndpoint(containerID, ep.Id); err != nil {
-			if isInfraContainer(sandbox) || hcsshim.ErrComputeSystemDoesNotExist != err {
+		if hcnEp == nil {
```

Comment:
maybe use an `else if` "statement" to avoid one level of indentation

btw the `hcnEp == nil` case means that the container runtime is docker correct?

Reply:
Yes, `hcnEp == nil` means the runtime is Docker.

I want to keep them in same level.
```
		if hcnEp == nil {
                     // docker ops
		} else {
                    // containerd ops
		}
```